pkg/cluster/loadbalancerprofile.go

```diff
@@ -0,0 +1,396 @@
+
+	result, err := m.publicIPAddresses.List(ctx, resourceGroupName)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < len(result); i++ {
+		// <infraID>-pip-v4 is not necessarily managed but is the default installed outbound IP
+		if *result[i].Name == infraID+"-pip-v4" || strings.Contains(*result[i].Name, "-outbound-pip-v4") {
+			ipAddresses[*result[i].Name] = result[i]
+		}
+	}
+
+	return ipAddresses, err
+}
+
+func genManagedOutboundIPName() string {
+	return uuid.DefaultGenerator.Generate() + "-outbound-pip-v4"
+}
+
+// Create a managed outbound IP Address.
+func (m *manager) createPublicIPAddress(ctx context.Context) (mgmtnetwork.PublicIPAddress, error) {
+	name := genManagedOutboundIPName()
+	resourceGroupName := stringutils.LastTokenByte(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, '/')
+	resourceID := fmt.Sprintf("%s/providers/Microsoft.Network/publicIPAddresses/%s", m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, name)
+	m.log.Infof("creating public IP Address: %s", name)
+	publicIPAddress := mgmtnetwork.PublicIPAddress{
+		Name:     &name,
+		ID:       &resourceID,
+		Location: &m.doc.OpenShiftCluster.Location,
+		PublicIPAddressPropertiesFormat: &mgmtnetwork.PublicIPAddressPropertiesFormat{
+			PublicIPAllocationMethod: mgmtnetwork.Static,
+			PublicIPAddressVersion:   mgmtnetwork.IPv4,
+		},
+		Sku: &mgmtnetwork.PublicIPAddressSku{
+			Name: mgmtnetwork.PublicIPAddressSkuNameStandard,
+		},
+	}
+
+	err := m.publicIPAddresses.CreateOrUpdateAndWait(ctx, resourceGroupName, name, publicIPAddress)
+	if err != nil {
+		return mgmtnetwork.PublicIPAddress{}, err
+	}
+
+	return publicIPAddress, nil
+}
+
+func getOutboundIPsFromLB(lb mgmtnetwork.LoadBalancer) []api.ResourceReference {
+	var outboundIPs []api.ResourceReference
+	fipConfigs := make(map[string]mgmtnetwork.FrontendIPConfiguration, len(*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations))
+
+	for i := 0; i < len(*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations); i++ {
+		fipConfigID := *(*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations)[i].ID
+		fipConfig := (*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations)[i]
+		fipConfigs[fipConfigID] = fipConfig
+	}
+
+	for _, obRule := range *lb.LoadBalancerPropertiesFormat.OutboundRules {
+		if *obRule.Name == outboundRuleV4 {
+			for i := 0; i < len(*obRule.OutboundRulePropertiesFormat.FrontendIPConfigurations); i++ {
+				id := *(*obRule.OutboundRulePropertiesFormat.FrontendIPConfigurations)[i].ID
+				if fipConfig, ok := fipConfigs[id]; ok {
+					outboundIPs = append(outboundIPs, api.ResourceReference{ID: *fipConfig.PublicIPAddress.ID})
+				}
+			}
+		}
+	}
+
+	return outboundIPs
+}
+
+func (m *manager) patchEffectiveOutboundIPs(ctx context.Context, outboundIPs []api.ResourceReference) error {
+	m.log.Info("patching effectiveOutboundIPs")
+	effectiveOutboundIPs := make([]api.EffectiveOutboundIP, 0, len(outboundIPs))
```

Comment:
I noticed in line 56, we check if `effectiveOutboundIPs` is nil. If we use `make` here, this will never be nil, so is the nil case only for new clusters? If a cluster has no effective outbound IPs, we just represent this as an empty slice?

Reply:
The nil case is only for new clusters.  If a cluster has a outboundType of `Loadbalancer` there will always be at least 1 effective outbound IP.  A cluster with outboundType of `UserDefinedRouting` will not have a `LoadBalancerProfile` thus no `effectiveOutboundIPs`.